frontend/src/cljs/imas_seamap/map/utils.cljs

```diff
@@ -416,34 +416,131 @@
+    filter-combinations))
+
+(defn- ->control [{:keys [cql-property] :as control} {:keys [id controls] :as rich-layer} db]
+  (let [values (get-in db [:map :rich-layers :async-datas id :controls cql-property :values])
+        value  (control->value control rich-layer db)
+        other-controls
+        (->>
+         controls
+         (remove #(= cql-property (:cql-property %)))
+         (map #(assoc % :value (control->value % rich-layer db))))
+        filter-combinations (get-in db [:map :rich-layers :async-datas id :filter-combinations])
+        valid-filter-combinations (reduce #(remove-incompatible-combinations %1 %2) filter-combinations other-controls)
+        valid-values (set (map #(get % cql-property) valid-filter-combinations))]
+    (assoc
+     control
+     :values (mapv #(hash-map :value % :valid? (boolean (some #{%} valid-values))) values)
+     :value  value
+     :is-default-value? (control-is-default-value? control rich-layer db))))
+
 (defn enhance-rich-layer
   "Takes a rich-layer and enhances the info with other layer data."
-  [{:keys [slider-label alternate-views timeline]
-    alternate-views-selected-id :alternate-views-selected
-    timeline-selected-id :timeline-selected
-    :as rich-layer}
-   rich-layers]
-  (let [alternate-views-selected  (first-where #(= (get-in % [:layer :id]) alternate-views-selected-id) alternate-views)
-        alternate-view-rich-layer (get rich-layers alternate-views-selected-id)
-        timeline                  (or (:timeline alternate-view-rich-layer) timeline)
+  [{:keys [id slider-label alternate-views timeline controls]
+    :as rich-layer} db]
+  (let [{:keys [tab]
+         alternate-views-selected-id :alternate-views-selected
+         timeline-selected-id        :timeline-selected
+         :as state}
+        (get-in db [:map :rich-layers :states id])
+        async-data                (get-in db [:map :rich-layers :async-datas id])
+
+        alternate-views              (mapv #(->alternate-view % db) alternate-views)
+        alternate-views-selected     (first-where #(= (get-in % [:layer :id]) alternate-views-selected-id) alternate-views)
+        alternate-view-rich-layer-id (get-in db [:map :rich-layers :layer-lookup alternate-views-selected-id])
+        alternate-view-rich-layer    (first-where #(= (:id %) alternate-view-rich-layer-id) (get-in db [:map :rich-layers :rich-layers]))
+
+        timeline                  (mapv #(->timeline % db) (or (:timeline alternate-view-rich-layer) timeline))
+        timeline-selected         (first-where #(= (get-in % [:layer :id]) timeline-selected-id) timeline)
         slider-label              (or (:slider-label alternate-view-rich-layer) slider-label)
-        timeline-selected         (first-where #(= (get-in % [:layer :id]) timeline-selected-id) timeline)]
+
+        controls                  (mapv #(->control % rich-layer db) controls)
+        cql-filter                (apply
```

Comment:
This sort of thing inline definitely makes me think we could/should have some CQL utilities :)

Reply:
Yep, exactly. Working on it now, actually (since I need them for the smart pill region selection too now)